scertecd/scertecd.go

```diff
@@ -0,0 +1,862 @@
+// foreground mode via its UpdateAll method.
+//
+// All exported fields must be initialized before calling an exported
+// method on the Server: either UpdateAll or Start.
+type Server struct {
+	SetecClient setec.Client                     // required client for setec
+	Domains     []string                         // domains to maintain certs for
+	Now         func() time.Time                 // if nil, time.Now is used
+	ACMEContact string                           // optional email address for ACME registration
+	Prefix      string                           // setec secret prefix
+	Logf        func(format string, args ...any) // alteranate log function; if nil, log.Printf is used
+
+	dtsOnce sync.Once
+	dts     []domainAndType
+
+	mu       sync.Mutex
+	acLazy   *acme.Client // nil until needed via getACMEClient
+	last     map[domainAndType]*certUpdateCheck
+	secCache map[string]*api.SecretValue
+}
+
+func (s *Server) initDomainAndTypes() {
+	s.dtsOnce.Do(func() {
+		for _, d := range s.Domains {
+			for _, typ := range []CertType{RSACert, ECDSACert} {
+				s.dts = append(s.dts, domainAndType{d, typ})
+			}
+		}
+	})
+}
+
+func (s *Server) logf(format string, args ...any) {
+	if s.Logf != nil {
+		s.Logf(format, args...)
+	} else {
+		log.Printf(format, args...)
+	}
+}
+
+// UpdateAll checks or updates all certs once and returns.
+//
+// If all certs are either fine or successfully updated, it returns nil.
+//
+// It is not necessary to call Start before UpdateAll.
+func (s *Server) UpdateAll() error {
+	s.initDomainAndTypes()
+	for _, dt := range s.dts {
+		cu := s.newCertUpdateCheck(dt)
+		st, err := cu.updateIfNeeded(context.Background())
+		if err != nil {
+			cu.lg.Printf("updateIfNeeded error: %v", err)
+			return err
+		}
+		cu.lg.Printf("success; %+v", st)
+	}
+	return nil
+}
+
+// Start starts a background renewal goroutine for each cert domain and
+// algorithm type. The context is used only for the initial ACME registration
+// check and not used thereafter.
+func (s *Server) Start(ctx context.Context) error {
+	s.initDomainAndTypes()
+	if _, err := s.getACMEClient(ctx); err != nil {
+		return err
+	}
+	for _, dt := range s.dts {
+		go s.renewCertLoop(dt)
+	}
+	return nil
+}
+
+func (s *Server) getACMEClient(ctx context.Context) (*acme.Client, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.acLazy != nil {
+		return s.acLazy, nil
+	}
+
+	ac, err := s.getOrMakeACMEClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getOrMakeACMEClient: %w", err)
+	}
+	if err := s.initACMEReg(ctx, ac); err != nil {
+		return nil, fmt.Errorf("initACMEReg: %w", err)
+	}
+	s.acLazy = ac
+	return ac, nil
+}
+
+var tmpls = template.Must(template.New("root").Parse(`
+{{define "root"}}
+<html><h1>scertecd</h1><table border=1 cellpadding=5>
+{{range .Certs}}
+   <tr>
+       <td><b>{{.Name}}</b>
+	   <td>{{.Status}}
+	       {{if .Log}}<pre>{{.Log}}</pre>{{end}}
+	   </td>
+   </tr>
+{{end}}
+</table></html>
+{{end}}
```

Comment:
you don't have to define a template, you can simply `tmpls.Execute(w, data)` and it will grab the first one
```suggestion
<html><h1>scertecd</h1><table border=1 cellpadding=5>
{{range .Certs}}
   <tr>
       <td><b>{{.Name}}</b>
	   <td>{{.Status}}
	       {{if .Log}}<pre>{{.Log}}</pre>{{end}}
	   </td>
   </tr>
{{end}}
</table></html>
```